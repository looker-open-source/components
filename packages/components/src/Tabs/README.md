# Tabs

_As compared with Tabs2, Tabs is the more composable version and should be used when custom behavior is required, e.g. icon actions within the tab button._

`Tabs` expects to receive a `TabList` with one or more `Tab` components. `TabPanels` expects `TabPanel`(s). `Tabs` manages the visibility of a single `TabPanel` based on its internal state.

The order in which elements are placed inside `TabList` and `TabPanels` defines the relationship between them. The first `Tab` is related to the first `TabPanel` and so on.

```tsx
function Basic() {
  const tabs = new Array(3).fill('tab');
  const tabPrefix = 'My Awesome Tab';
  return (
    <Tabs>
      <TabList>
        {tabs.map((_k, index) => (
          <Tab key={index}>
            {tabPrefix} {index}
          </Tab>
        ))}
      </TabList>
      <TabPanels>
        {tabs.map((_k, index) => (
          <TabPanel key={index}>This is {index}</TabPanel>
        ))}
      </TabPanels>
    </Tabs>
  );
}
```

## Horizonal Scroll

`Tabs` supports scrolling when the tabs' combined width cannot be accomdated within the viewport width. - Try scrolling left-right.

```tsx
function Basic() {
  const tabs = new Array(20).fill('tab');
  const tabPrefix = 'My Awesome Tab';
  return (
    <Tabs>
      <TabList>
        {tabs.map((_k, index) => (
          <Tab key={index}>
            {tabPrefix} {index}
          </Tab>
        ))}
      </TabList>
      <TabPanels>
        {tabs.map((_k, index) => (
          <TabPanel key={index}>This is {index}</TabPanel>
        ))}
      </TabPanels>
    </Tabs>
  );
}
```

## Disabled Tab(s)

To create a disabled Tab add the `disabled` prop. NOTE: A `TabPanel` must still exist for disabled tabs to allow for tabs to keep the index positions in sync.

## Distribute Attribute

Using the distribute prop on `TabList` will equally distributed each `Tab` in the container.

```tsx
function Distributed() {
  const tabs = new Array(3).fill('tab');
  const tabPrefix = 'My Awesome Tab';
  return (
    <Tabs>
      <TabList distribute>
        {tabs.map((_k, index) => (
          <Tab key={index}>
            {tabPrefix} {index}
          </Tab>
        ))}
      </TabList>
      <TabPanels>
        {tabs.map((_k, index) => (
          <TabPanel key={index}>This is {index}</TabPanel>
        ))}
      </TabPanels>
    </Tabs>
  );
}
```

## Controlled Tab

```tsx
function Distributed() {
  const [currentTabIndex, setTab] = useState(0);
  return (
    <>
      <Space>
        <Button onClick={() => setTab(0)}>Go to A</Button>
        <Button onClick={() => setTab(1)}>Go to B</Button>
      </Space>
      <Tabs index={currentTabIndex} onChange={setTab}>
        <TabList>
          <Tab>A</Tab>
          <Tab>B</Tab>
        </TabList>
        <TabPanels>
          <TabPanel>A</TabPanel>
          <TabPanel>B</TabPanel>
        </TabPanels>
      </Tabs>
    </>
  );
}
```
