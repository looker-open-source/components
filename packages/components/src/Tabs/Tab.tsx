/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { forwardRef, Ref, useState } from 'react'
import { rgba } from 'polished'
import styled from 'styled-components'
import {
  border,
  BorderProps,
  CompatibleHTMLProps,
  layout,
  LayoutProps,
  reset,
  space,
  SpaceProps,
  typography,
  TypographyProps,
} from '@looker/design-tokens'

export interface TabProps
  extends Omit<CompatibleHTMLProps<HTMLButtonElement>, 'type'>,
    BorderProps,
    LayoutProps,
    SpaceProps,
    TypographyProps {
  disabled?: boolean
  focusVisible?: boolean
  selected?: boolean
  onSelect?: () => void
}

const TabStyle = styled.button<TabProps>`
  ${reset}
  border: none;

  ${border}
  ${layout}
  ${space}
  ${typography}

  background: transparent;
  border-bottom: 3px solid;
  border-bottom-color: ${(props) =>
    props.selected ? props.theme.colors.palette.purple400 : 'transparent'};
  border-radius: 0;
  color: ${(props) =>
    props.selected
      ? props.theme.colors.palette.charcoal800
      : props.theme.colors.palette.charcoal500};
  cursor: pointer;

  & + & {
    margin-left: ${(props) => props.theme.space.xlarge};
  }

  &:active {
    border-bottom-color: ${(props) =>
      props.selected
        ? props.theme.colors.palette.purple400
        : props.theme.colors.palette.charcoal400};
  }

  &:active,
  &:hover {
    border-bottom-color: transparent;
  }

  &:focus {
    outline: none;
  }

  ${(props) =>
    props.focusVisible &&
    `box-shadow: 0 0 0 0.15rem ${rgba(
      props.theme.colors.palette.purple300,
      0.25
    )};`};

  &:hover {
    border-bottom-color: ${(props) =>
      props.selected
        ? props.theme.colors.palette.purple400
        : props.theme.colors.palette.charcoal300};
  }

  &:disabled {
    cursor: default;
    color: ${(props) => props.theme.colors.palette.charcoal300};
    border-bottom-color: transparent;
  }
`

const TabJSX = forwardRef((props: TabProps, ref: Ref<HTMLButtonElement>) => {
  const { children, disabled, onBlur, onKeyUp, onSelect, ...restProps } = props

  const [isFocusVisible, setFocusVisible] = useState(false)

  const handleOnKeyUp = (event: React.KeyboardEvent<HTMLButtonElement>) => {
    setFocusVisible(true)
    onKeyUp && onKeyUp(event)
  }

  const handleOnBlur = (event: React.FocusEvent<HTMLButtonElement>) => {
    setFocusVisible(false)
    onBlur && onBlur(event)
  }

  const onClick = () => {
    if (!disabled && onSelect) {
      onSelect()
    }
    setFocusVisible(false)
  }

  return (
    <TabStyle
      focusVisible={isFocusVisible}
      onKeyUp={handleOnKeyUp}
      onBlur={handleOnBlur}
      onClick={onClick}
      disabled={disabled}
      {...restProps}
      ref={ref}
    >
      {children}
    </TabStyle>
  )
})

TabJSX.displayName = 'TabJSX'

export const Tab = styled(TabJSX)``

Tab.defaultProps = {
  fontSize: 'small',
  fontWeight: 'semiBold',
  minWidth: '3rem',
  pb: 'small',
  pt: 'xsmall',
}
