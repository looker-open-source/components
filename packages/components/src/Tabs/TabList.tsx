/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, {
  Children,
  cloneElement,
  forwardRef,
  KeyboardEvent,
  useRef,
  Ref,
} from 'react'
import { SpaceProps, space } from '@looker/design-tokens'
import styled, { css } from 'styled-components'
import { fontSize, FontSizeProps } from 'styled-system'
import { moveFocus, useForkedRef } from '../utils'
import { TabContext } from './TabContext'
import { Tab } from '.'

export interface TabListProps extends SpaceProps, FontSizeProps {
  children: JSX.Element[]
  selectedIndex?: number
  onSelectTab?: (index: any) => void
  className?: string
  distribute?: boolean
}

const TabListLayout = forwardRef(
  (
    { children, selectedIndex, onSelectTab, className }: TabListProps,
    ref: Ref<HTMLDivElement>
  ) => {
    const wrapperRef = useRef<HTMLDivElement>(null)
    const forkedRef = useForkedRef(wrapperRef, ref)

    const clonedChildren = Children.map(
      children,
      (child: JSX.Element, index: number) => {
        return cloneElement(child, {
          index,
          onSelect: () => onSelectTab && onSelectTab(index),
          selected: index === selectedIndex,
          selectedIndex,
        })
      }
    )

    function handleArrowKey(direction: number, initial: number) {
      moveFocus(direction, initial, wrapperRef)
    }

    const context = {
      handleArrowLeft: (e: KeyboardEvent<HTMLButtonElement>) => {
        e.preventDefault()
        handleArrowKey(-1, -1)
        return false
      },
      handleArrowRight: (e: KeyboardEvent<HTMLButtonElement>) => {
        e.preventDefault()
        handleArrowKey(1, 0)
        return false
      },
    }

    return (
      <TabContext.Provider value={context}>
        <div
          aria-label="Tabs"
          className={className}
          ref={forkedRef}
          role="tablist"
        >
          {clonedChildren}
        </div>
      </TabContext.Provider>
    )
  }
)

TabListLayout.displayName = 'TabListLayout'

const defaultLayoutCSS = css`
  ${Tab} {
    min-width: 3rem;
  }

  ${Tab} + ${Tab} {
    margin-left: ${(props) => props.theme.space.xlarge};
  }
`

const distributeCSS = css`
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;

  ${Tab} {
    padding: ${({ theme: { space } }) => `${space.xsmall} ${space.medium}`};
  }
`

export const TabList = styled(TabListLayout)`
  ${space}
  ${fontSize}

  border-bottom: 1px solid ${(props) => props.theme.colors.ui2};
  -ms-overflow-style: none; /* Internet Explorer 10+ */
  overflow-x: auto;
  scrollbar-width: none; /* Firefox */
  white-space: nowrap;

  &::-webkit-scrollbar {
    display: none; /* Safari and Chrome */
  }

  ${({ distribute }) => (distribute ? distributeCSS : defaultLayoutCSS)}
`

TabList.defaultProps = {
  fontSize: 'small',
}
