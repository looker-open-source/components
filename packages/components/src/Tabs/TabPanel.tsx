/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import type { FC } from 'react'
import React from 'react'
import styled from 'styled-components'

export interface TabPanelProps {
  className?: string
  selected?: boolean
  /**
   * Set to `true` if you would like TabPanel to be reached via tab-key.
   * Generally this is _only_ the case when the TabPanel contains no tab-stopping items (a, button, etc.)
   * @default false
   */
  isTabStop?: boolean
}

const TabPanelLayout: FC<TabPanelProps> = ({
  children,
  className,
  selected,
  isTabStop = false,
}) =>
  selected ? (
    <div className={className} tabIndex={isTabStop ? 0 : -1}>
      {children}
    </div>
  ) : null

/**
 * @deprecated Use `Tabs2` & `Tab2` instead
 */
export const TabPanel = styled(TabPanelLayout)`
  height: 100%;
  outline: none;
`
