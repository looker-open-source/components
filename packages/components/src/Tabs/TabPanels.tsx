/**
 * Copyright (c) 2023 Google LLC
 * SPDX-License-Identifier: MIT
 */

import React, { Children, cloneElement } from 'react';
import styled from 'styled-components';
import type {
  FlexboxProps,
  LayoutProps,
  SpaceProps,
} from '@looker/design-tokens';
import { flexbox, layout, space, reset } from '@looker/design-tokens';

export interface TabPanelsProps extends FlexboxProps, LayoutProps, SpaceProps {
  children: JSX.Element | JSX.Element[];
  className?: string;
  selectedIndex?: number;
  onSelectTab?: (index: number) => void;
}

const Layout = ({
  children,
  className,
  selectedIndex,
  ...props
}: TabPanelsProps) => {
  const { onSelectTab: _onSelectTab, ...tabPanelsLayoutProps } = props;

  const clonedChildren = Children.map(
    children,
    (child: JSX.Element, index: number) => {
      return cloneElement(child, {
        selected: index === selectedIndex,
      });
    }
  );

  return (
    <div
      aria-labelledby={`tab-${selectedIndex}`}
      className={className}
      id={`panel-${selectedIndex}`}
      role="tabpanel"
      {...tabPanelsLayoutProps}
    >
      {clonedChildren}
    </div>
  );
};

export const TabPanels = styled(Layout).attrs(({ pt = 'large' }) => ({
  pt,
}))`
  ${reset}
  ${flexbox}
  ${layout}
  ${space}
`;
