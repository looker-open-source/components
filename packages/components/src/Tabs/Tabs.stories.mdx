import { Code, MessageBar, Tabs2 } from '@looker/components'
import { Meta } from '@storybook/addon-docs'

<Meta component={Tabs2} title="Docs/Components/Content/Tabs" />

# Tabs

<MessageBar noActions intent="inform">
  Tabs has been deprecated and replaced with
  <Code fontWeight="bold" fontSize="inherit">
    Tabs2
  </Code> please update any existing instances to use the new improved version. Deprecation
  of `Tabs` is planned for @looker/components 3.0
</MessageBar>

`Tabs2` contains one or more `Tab2` components. `Tabs2`, when left uncontrolled, manages the visible tab as well as switching between tabs.

Each `Tab2` must have a `label`. The value passed as children will be displayed as its content area. `id` is optional (`label` will be used in its place if not supplied) but `id` is highly recommended when using `Tabs2` in a controlled state.

```jsx
<Tabs2>
  <Tab2 id="cats" label="Cats">
    Here's awesome story about cats
  </Tab2>
  <Tab2 id="dogs" label="Dogs">
    Cats are way better than dogs. Go to other tab
  </Tab2>
  <Tab2 label="Fish">Are kinda smelly</Tab2>
</Tabs2>
```

## Horizonal Scroll

`Tabs` supports scrolling when the tabs' combined width cannot be accommodated within the viewport width. Try scrolling left-right.

```jsx
<Tabs2>
  <Tab2 label="Really long tab title that should help demonstrate side-scroll">
    Really long tab title that should help demonstrate side-scroll
  </Tab2>
  <Tab2 label="Yet-another-tab-title to get things wide enough">
    Yet-another-tab-title to get things wide enough
  </Tab2>
  <Tab2 label="Really, you must be obsessed with these long tab titles?">
    Really, you must be obsessed with these long tab titles?
  </Tab2>
  <Tab2 label="Okay, okay, we get it you want to see scrolling in action...">
    Okay, okay, we get it you want to see scrolling in action...
  </Tab2>
  <Tab2 label="Repeat really long tab title that should help demonstrate side-scroll">
    Repeat really long tab title that should help demonstrate side-scroll
  </Tab2>
  <Tab2 label="Repeat yet-another-tab-title to get things wide enough">
    Repeat yet-another-tab-title to get things wide enough
  </Tab2>
</Tabs2>
```

## Disabled Tab(s)

To display a tab that can't be activated add the `disabled` prop.

```jsx
<Tabs2>
  <Tab2 id="cats" label="Cats">
    Here's awesome story about cats
  </Tab2>
  <Tab2 id="dogs" label="Dogs">
    Cats are way better than dogs. Go to other tab
  </Tab2>
  <Tab2 label="Fish">Are kinda smelly</Tab2>
  <Tab2 disabled label="Human">
    not available
  </Tab2>
</Tabs2>
```

## Distribute Attribute

Using the distribute prop on `Tabs2` will equally distribute each `Tab` in the container.

```jsx
<Tabs2 distribute>
  <Tab2 label="Rebecca Purple">
    Rebecca Purple
    <div
      style={{
        backgroundColor: 'RebeccaPurple',
        height: '350px',
      }}
    />
  </Tab2>
  <Tab2 label="Coral">
    <div
      style={{
        backgroundColor: 'coral',
        height: '350px',
      }}
    />
  </Tab2>
  <Tab2 label="Forest Green">
    <div
      style={{
        backgroundColor: 'ForestGreen',
        height: '350px',
      }}
    />
  </Tab2>
  <Tab2 label="Cornflower Blue">
    <div
      style={{
        backgroundColor: 'CornflowerBlue',
        height: '350px',
      }}
    />
  </Tab2>
</Tabs2>
```

## Controlled Tab

```jsx
;() => {
  const [tabId, setTabId] = useState('cats')
  return (
    <>
      <button onClick={() => setTabId('cats')}>Go to Cats</button>
      <button onClick={() => setTabId('dogs')}>Go to Dogs</button>
      <Tabs2 tabId={tabId} onTabChange={setTabId}>
        <Tab2 id="cats" label="Cats">
          Here's awesome story about cats
        </Tab2>
        <Tab2 id="dogs" label="Dogs">
          Cats are way better than dogs. Go to other tab
        </Tab2>
      </Tabs2>
    </>
  )
}
```
