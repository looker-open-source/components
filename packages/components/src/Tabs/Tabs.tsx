/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { Children, cloneElement, FC, useRef, useState } from 'react'

export interface UseTabsProps {
  controlledIndex?: number
  defaultIndex?: number
  index?: number
  isControlled?: boolean
  onChange?: (index: number) => void
}

export interface TabsProps extends UseTabsProps {
  children: JSX.Element[]
}

export function useTabs(props?: UseTabsProps) {
  const defaultIndex = (props && props.defaultIndex) || 0
  const onChange = props && props.onChange
  const isControlled = (props && props.isControlled) || false

  const [selectedIndex, setSelectedIndex] = useState(defaultIndex)

  return {
    onSelectTab: (index: number) => {
      onChange && onChange(index)
      if (!isControlled) {
        setSelectedIndex(index)
      }
    },
    selectedIndex:
      isControlled && props ? props.controlledIndex : selectedIndex,
  }
}

export const Tabs: FC<TabsProps> = ({
  children,
  index: controlledIndex,
  defaultIndex,
  onChange,
}) => {
  const { current: isControlled } = useRef(controlledIndex !== undefined)

  /* eslint-disable no-console */
  if (!isControlled && controlledIndex !== undefined) {
    console.warn(
      'Tabs is changing from uncontrolled to controlled. Tabs should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled Tabs for the lifetime of the component. Check the `index` prop being passed in.'
    )
  }

  if (isControlled && controlledIndex === undefined) {
    console.warn(
      'Tabs is changing from uncontrolled to controlled. Tabs should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled Tabs for the lifetime of the component. Check the `index` prop being passed in.'
    )
  }
  /* eslint-enable no-console */

  const tabs = useTabs({
    controlledIndex,
    defaultIndex,
    isControlled,
    onChange,
  })

  const clonedChildren = Children.map(children, (child: JSX.Element) => {
    return cloneElement(child, tabs)
  })

  return <>{clonedChildren}</>
}
