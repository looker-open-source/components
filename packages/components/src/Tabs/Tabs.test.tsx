/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import '@testing-library/jest-dom/extend-expect'
import {
  assertSnapshotShallow,
  mountWithTheme,
  renderWithTheme,
  shallowWithTheme,
} from '@looker/components-test-utils'
import React from 'react'
import { fireEvent } from '@testing-library/react'
import { Tab } from './Tab'
import { TabList } from './TabList'
import { TabPanel } from './TabPanel'
import { TabPanels } from './TabPanels'
import { Tabs, useTabs } from './Tabs'

test('Tabs snapshot works as expected', () => {
  assertSnapshotShallow(
    <Tabs>
      <TabList>
        <Tab>tab1</Tab>
        <Tab>tab2</Tab>
        <Tab disabled>tab3</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>this is tab1 content</TabPanel>
        <TabPanel>this is tab2 content</TabPanel>
        <TabPanel>this is tab3 content</TabPanel>
      </TabPanels>
    </Tabs>
  )
})

test('shows the correct number of navigation tabs', () => {
  const tabs = shallowWithTheme(
    <Tabs>
      <TabList>
        <Tab>tab1</Tab>
        <Tab>tab2</Tab>
        <Tab disabled>tab3</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>this is tab1 content</TabPanel>
        <TabPanel>this is tab2 content</TabPanel>
        <TabPanel>this is tab3 content</TabPanel>
      </TabPanels>
    </Tabs>
  )
  expect(tabs.find(Tab)).toHaveLength(3)
})

test('starts with Tab at index 0 opened', () => {
  const tabs = mountWithTheme(
    <Tabs>
      <TabList>
        <Tab>tab1</Tab>
        <Tab>tab2</Tab>
        <Tab disabled>tab3</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>this is tab1 content</TabPanel>
        <TabPanel>this is tab2 content</TabPanel>
        <TabPanel>this is tab3 content</TabPanel>
      </TabPanels>
    </Tabs>
  )

  const EachTabPanel: any = tabs
    .find(TabPanels)
    .find(TabPanel)
    .map((panel: any) => panel.props())

  expect(EachTabPanel[0].children).toEqual('this is tab1 content')
  expect(EachTabPanel[0].selected).toEqual(true)
  expect(EachTabPanel[1].selected).toEqual(false)
  expect(EachTabPanel[2].selected).toEqual(false)
})

const ChangingPanel = () => {
  return (
    <Tabs>
      <TabList>
        <Tab>tab1</Tab>
        <Tab>tab2</Tab>
        <Tab disabled>tab3</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>this is tab1 content</TabPanel>
        <TabPanel>this is tab2 content</TabPanel>
        <TabPanel>this is tab3 content</TabPanel>
      </TabPanels>
    </Tabs>
  )
}

test('clicking on tab opens correct panel', () => {
  const { getByText, queryByText } = renderWithTheme(<ChangingPanel />)

  expect(queryByText('this is tab1 content')).toBeInTheDocument()
  expect(queryByText('this is tab2 content')).not.toBeInTheDocument()
  fireEvent.click(getByText('tab2'))
  expect(queryByText('this is tab1 content')).not.toBeInTheDocument()
  expect(queryByText('this is tab2 content')).toBeInTheDocument()
})

const DisableTab = () => {
  return (
    <Tabs>
      <TabList>
        <Tab>tab1</Tab>
        <Tab>tab2</Tab>
        <Tab disabled>tab3</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>this is tab1 content</TabPanel>
        <TabPanel>this is tab2 content</TabPanel>
        <TabPanel>this is the disable tab-panel</TabPanel>
      </TabPanels>
    </Tabs>
  )
}

test('clicking on disable tab does not change panel', () => {
  const { getByText, queryByText } = renderWithTheme(<DisableTab />)

  expect(queryByText('this is tab1 content')).toBeInTheDocument()
  expect(queryByText('this is the disable tab-panel')).not.toBeInTheDocument()
  fireEvent.click(getByText('tab3'))
  expect(queryByText('this is tab1 content')).toBeInTheDocument()
  expect(queryByText('this is the disable tab-panel')).not.toBeInTheDocument()
})

const TabHooks = () => {
  const tab = useTabs()
  return (
    <>
      <TabList {...tab}>
        <Tab>Tab Hook 1</Tab>
        <Tab>Tab Hook 2</Tab>
      </TabList>
      <TabPanels {...tab}>
        <TabPanel>1 this is the panel of tab hook 1</TabPanel>
        <TabPanel>2 this is the panel of tab hook 2</TabPanel>
      </TabPanels>
    </>
  )
}

test('hooks working', () => {
  const { getByText, queryByText } = renderWithTheme(<TabHooks />)

  expect(queryByText('1 this is the panel of tab hook 1')).toBeInTheDocument()
  expect(
    queryByText('2 this is the panel of tab hook 2')
  ).not.toBeInTheDocument()
  fireEvent.click(getByText('Tab Hook 2'))
  expect(
    queryByText('1 this is the panel of tab hook 1')
  ).not.toBeInTheDocument()
  expect(queryByText('2 this is the panel of tab hook 2')).toBeInTheDocument()
})

describe('focus behavior', () => {
  const TabTest = () => (
    <Tabs>
      <TabList>
        <Tab>tab1</Tab>
        <Tab>tab2</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>this is tab1 content</TabPanel>
        <TabPanel>this is tab2 content</TabPanel>
      </TabPanels>
    </Tabs>
  )

  test('Tab Focus: does not render focus ring after click', () => {
    const { getByText } = renderWithTheme(<TabTest />)

    fireEvent.click(getByText('tab1'))
    expect(getByText('tab1')).toMatchSnapshot()
  })

  test('Tab Focus: renders focus ring for keyboard navigation', () => {
    const { getByText } = renderWithTheme(<TabTest />)

    fireEvent.keyUp(getByText('tab2'), { charCode: 9, code: 9, key: 'Tab' })
    expect(getByText('tab2')).toMatchSnapshot()
  })
})
