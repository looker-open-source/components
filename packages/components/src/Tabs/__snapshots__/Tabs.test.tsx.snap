// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs snapshot works as expected 1`] = `ShallowWrapper {}`;

exports[`focus behavior Tab Focus: does not render focus ring after click 1`] = `
.c1 {
  min-width: 3rem;
  padding-bottom: 0.75rem;
  padding-top: 0.5rem;
  font-size: 0.875rem;
  font-weight: 600;
  background: transparent;
  border: none;
  border-bottom: 3px solid;
  border-bottom-color: #6C43E0;
  border-radius: 0;
  color: #262D33;
  cursor: pointer;
}

.c1 + .c0 {
  margin-left: 2rem;
}

.c1:active {
  border-bottom-color: #6C43E0;
}

.c1:active,
.c1:hover {
  border-bottom-color: transparent;
}

.c1:focus {
  outline: none;
}

.c1:hover {
  border-bottom-color: #6C43E0;
}

.c1:disabled {
  border-bottom-color: transparent;
  color: #C1C6CC;
  cursor: default;
}

.c2 + .c0 {
  margin-left: 2rem;
}

<button
  class="c0 c1 "
  font-size="small"
  font-weight="semiBold"
>
  tab1
</button>
`;

exports[`focus behavior Tab Focus: renders focus ring for keyboard navigation 1`] = `
.c2 + .c0 {
  margin-left: 2rem;
}

.c3 + .c0 {
  margin-left: 2rem;
}

.c1 {
  min-width: 3rem;
  padding-bottom: 0.75rem;
  padding-top: 0.5rem;
  font-size: 0.875rem;
  font-weight: 600;
  background: transparent;
  border: none;
  border-bottom: 3px solid;
  border-bottom-color: transparent;
  border-radius: 0;
  color: #707781;
  cursor: pointer;
  box-shadow: 0 0 0 0.15rem rgba(151,133,242,0.25);
}

.c1 + .c0 {
  margin-left: 2rem;
}

.c1:active {
  border-bottom-color: #707781;
}

.c1:active,
.c1:hover {
  border-bottom-color: transparent;
}

.c1:focus {
  outline: none;
}

.c1:hover {
  border-bottom-color: #C1C6CC;
}

.c1:disabled {
  border-bottom-color: transparent;
  color: #C1C6CC;
  cursor: default;
}

<button
  class="c0 c1 "
  font-size="small"
  font-weight="semiBold"
>
  tab2
</button>
`;
